app/classes/products_recommendation.rb

```diff
@@ -0,0 +1,128 @@
+module ProductsRecommendation
+  def self.filter_products_with_aggregation(user, limit = 5, skip = 0)
+    pipeline = []
+  
+    pipeline << {
+      '$match': { 'status': 'APPROVED' }
+    }
+  
+    pipeline << {
+      '$lookup': {
+        from: 'reviews',
+        localField: '_id',
+        foreignField: 'product_id',
+        as: 'reviews'
+      }
+    }
+  
+    pipeline << {
+      '$addFields': {
+        'total_reviews': { '$size': '$reviews' },
+        'positive_reviews': {
+          '$size': {
+            '$filter': {
+              'input': '$reviews',
+              'as': 'review',
+              'cond': '$$review.rating'
+            }
+          }
+        }
+      }
+    }
+  
+    pipeline << {
+      '$addFields': {
+        'positive_review_percentage': {
+          '$cond': {
+            'if': { '$eq': ['$total_reviews', 0] },
+            'then': -1,
+            'else': { '$multiply': [{ '$divide': ['$positive_reviews', '$total_reviews'] }, 100] }
+          }
+        }
+      }
+    }
+  
+    # pipeline << {
+    #   '$addFields': {
+    #     'allergen_penalty': {
+    #       '$cond': {
+    #         'if': {
+    #           '$anyElementTrue': {
+    #             '$map': {
+    #               'input': '$allergens',
+    #               'as': 'allergen',
+    #               'in': { '$in': ['$$allergen', user.allergens_ids] }
+    #             }
+    #           }
+    #         },
+    #         'then': 35,
+    #         'else': 0
+    #       }
+    #     }
+    #   }
+    # } if user.allergens_ids.present?
+
+    # dietary_preference_branches = []
+    # dietary_preference_branches << { "case": { "$and": [{ "$eq": ["$vegan", false] }] }, "then": 40 } if user.dietary_preferences == :VEGAN
+    # dietary_preference_branches << { "case": { "$and": [{ "$eq": ["$vegetarian", false] }] }, "then": 40 } if user.dietary_preferences == :VEGETARIAN
+  
+    # pipeline << {
+    #   "$addFields": {
+    #     "dietary_preference_penalty": {
+    #       "$switch": {
+    #         "branches": dietary_preference_branches,
+    #         "default": 0
+    #       }
+    #     }
+    #   }
+    # } if dietary_preference_branches.any?
```

Comment:
This comment could be removed since it's no longer the intended way of fetching the data